Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -159,17 +367,55 @@ void FieldValue::SwitchTo(const Type type) {
   // Not same type. Destruct old type first and then initialize new type.
   // Must call destructor explicitly for any non-POD type.
   switch (tag_) {
+    case Type::Timestamp:
+      timestamp_value_.~Timestamp();
+      break;
+    case Type::ServerTimestamp:
+      server_timestamp_value_.~ServerTimestamp();
+      break;
+    case Type::String:
+      string_value_.~basic_string();
+      break;
+    case Type::Blob:
+      blob_value_.~Blob();
+      break;
+    case Type::GeoPoint:
+      geo_point_value_.~GeoPoint();
+      break;
     case Type::Array:
       array_value_.~vector();
       break;
+    case Type::Object:
+      object_value_.~map();
+      break;
     default: {}  // The other types where there is nothing to worry about.
   }
   tag_ = type;
   // Must call constructor explicitly for any non-POD type to initialize.
   switch (tag_) {
+    case Type::Timestamp:
+      new (&timestamp_value_) Timestamp(0, 0);
```

Comment:
The timestamp constructor would throw on these arguments.

Perhaps we're better off just making Timestamp default constructable.

Reply:
Okay, Timestamp is going to be a public type so we need to take some care.

My intuition is that Timestamp should default construct with zero values and we should add a Timestamp::Now() factory function to actually get current time.

/cc @varconst-fb

Reply:
I'm in favor of having a default constructor that initializes the value to zero:
- Timestamp looks like it should have value semantics;
- I think it's more idiomatic for a default-constructed object to be in an "empty" state;
- it's consistent with what std::chrono::time_point does.

Reply:
We've done this in the past with "PORTING NOTE" in the comments of the class in question where there's a divergence.

https://github.com/firebase/firebase-js-sdk/search?utf8=%E2%9C%93&q=%22PORTING+NOTE%22&type=